pkg/provider/notebook_resource.go

```diff
@@ -0,0 +1,248 @@
+	var plan NotebookResourceModel
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get use case ID if specified
+	var useCaseID string
+	if !plan.UseCaseID.IsNull() {
+		useCaseID = plan.UseCaseID.ValueString()
+	}
+
+	// Read file
+	filePath := plan.FilePath.ValueString()
+	content, err := os.ReadFile(filePath)
+	if err != nil {
+		resp.Diagnostics.AddError("Error reading file", fmt.Sprintf("Unable to read file at %s: %s", filePath, err))
+		return
+	}
+
+	// Calculate file hash
+	hash := sha256.Sum256(content)
+	hashStr := hex.EncodeToString(hash[:])
+
+	fileName := filepath.Base(filePath)
+
+	traceAPICall("ImportNotebookFromFile")
+	importResp, err := r.provider.service.ImportNotebookFromFile(ctx, fileName, content, useCaseID)
+	if err != nil {
+		resp.Diagnostics.AddError("Error importing notebook", err.Error())
+		return
+	}
+
+	// Update model
+	plan.ID = types.StringValue(importResp.ID)
+	plan.Name = types.StringValue(importResp.Name)
+	plan.FileHash = types.StringValue(hashStr)
+
+	// Save model into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (r *NotebookResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state NotebookResourceModel
+
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	traceAPICall("GetNotebook")
+	notebook, err := r.provider.service.GetNotebook(ctx, state.ID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Error reading notebook", err.Error())
+		return
+	}
+
+	// Update resource state with the current state
+	state.ID = types.StringValue(notebook.ID)
+	state.Name = types.StringValue(notebook.Name)
+
+	// Store the UseCaseID if available.
+	if notebook.UseCaseID != "" {
+		state.UseCaseID = types.StringValue(notebook.UseCaseID)
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (r *NotebookResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan, state NotebookResourceModel
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	diags = req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Handle use case updates.
+	if !state.UseCaseID.IsNull() && plan.UseCaseID.IsNull() {
+		// Remove notebook from old use case if it exists and re-import as a classic notebook.
+		traceAPICall("DeleteNotebook")
+		err := r.provider.service.DeleteNotebook(ctx, state.ID.ValueString())
+		if err != nil {
+			resp.Diagnostics.AddError("Error deleting notebook", err.Error())
+			return
+		}
+		// Re-import the notebook.
+		filePath := plan.FilePath.ValueString()
+		content, err := os.ReadFile(filePath)
```

Comment:
We can also open the file for reading and then Copy() it's content during importing. 
A more performant way to deal with files of considerable sizes.

Reply:
But I'll end up copying it twice, no? Once to hash, once to send to the API?